### PR TITLE
Be/feature/ri 4257 sync recommendations

### PR DESCRIPTION
**Pull request by AmirAllayarovSofteq**
#RI-4257 - sync recommendations

### CODE REVIEW

File: redisinsight/api/src/modules/database-analysis/database-analysis.service.ts [hunk 1]
```diff
@@ -22,6 +23,7 @@ export class DatabaseAnalysisService {
     private readonly analyzer: DatabaseAnalyzer,
     private readonly databaseAnalysisProvider: DatabaseAnalysisProvider,
     private readonly scanner: KeysScanner,
+    private liveRecommendationService: DatabaseRecommendationService,
```
Comment (arthosofteq) [inline redisinsight/api/src/modules/database-analysis/database-analysis.service.ts:26]: Why do you have inconsistent naming?

File: redisinsight/api/src/modules/database-recommendation/database-recommendation.service.ts [hunk 2]
```diff
@@ -85,11 +86,31 @@ export class DatabaseRecommendationService {
...
     return this.databaseRecommendationsProvider.update(clientMetadata, id, dto);
   }
 
+  /**
+   * Sync db analysis recommendations and live recommendations
+   * @param clientMetadata
+   * @param recommendations
+   */
+  public async sync(clientMetadata: ClientMetadata, recommendations: Recommendation[]): Promise<void> {
+    const db = clientMetadata.db ?? (await this.databaseService.get(clientMetadata.databaseId))?.db ?? 0;
```
Comment (arthosofteq) [inline redisinsight/api/src/modules/database-recommendation/database-recommendation.service.ts:103]: I thought that we decided to rid of logical db in recommendations, no?

File: redisinsight/api/src/modules/database-recommendation/providers/database-recommendation.provider.ts [hunk 3]
```diff
@@ -141,6 +150,30 @@ export class DatabaseRecommendationProvider {
...
+   * @param liveRecommendations
+   */
+  async sync(
+    clientMetadata: ClientMetadata,
+    dbAnalysisRecommendations: Recommendation[],
+    liveRecommendations: DatabaseRecommendation[],
+  ): Promise<void> {
+    this.logger.log('Synchronization of recommendations');
+    try {
+      const newRecommendations = differenceBy(dbAnalysisRecommendations, liveRecommendations, 'name');
```
Comment (arthosofteq) [inline redisinsight/api/src/modules/database-recommendation/providers/database-recommendation.provider.ts:166]: Fine with me for now but I feel like previously we wanted to sync recommendations one-by-one.
To have generic mechanism to create new and ignore existing recommendations. the same way as we are determining recommendations. 
With this approach you might have several the same recommendations with concurrent sync invocations, no?
Do you have unique constraint on db level?

**Review by arthosofteq — APPROVED** [review]

File: redisinsight/api/src/utils/recommendation-helper.ts [hunk 4]
```diff
@@ -1,6 +1,15 @@
-import { REDISEARCH_MODULES } from 'src/constants';
+import { sortBy } from 'lodash';
+import { REDISEARCH_MODULES, REDIS_STACK, RECOMMENDATION_NAMES } from 'src/constants';
 import { AdditionalRedisModule } from 'src/modules/database/models/additional.redis.module';
 
 export const isRedisearchModule = (modules: AdditionalRedisModule[]): boolean => modules?.some(
   ({ name }) => REDISEARCH_MODULES.some((search) => name === search),
 );
+
+export const sortRecommendations = (recommendations: any[]) => sortBy(recommendations, [
```
[the remark this answers is not in the document]
Comment (AmirAllayarovSofteq) [reply redisinsight/api/src/utils/recommendation-helper.ts:9]: It is business requirement (recommendations should be in the same order as we have in db analysis

File: redisinsight/api/src/modules/database-recommendation/providers/database-recommendation.provider.ts [hunk 5]
```diff
@@ -141,6 +150,30 @@ export class DatabaseRecommendationProvider {
...
+      const sortedRecommendations = sortRecommendations(newRecommendations);
```
[the remark this answers is not in the document]
Comment (AmirAllayarovSofteq) [reply redisinsight/api/src/modules/database-recommendation/providers/database-recommendation.provider.ts:167]: It is business requirement (recommendations should be in the same order as we have in db analysis

File: redisinsight/api/src/modules/database-analysis/database-analysis.service.ts (hunk 1 above)
Comment (AmirAllayarovSofteq) [reply redisinsight/api/src/modules/database-analysis/database-analysis.service.ts:26]: Renamed, thank you

File: redisinsight/api/src/modules/database-recommendation/providers/database-recommendation.provider.ts (hunk 3 above)
Comment (AmirAllayarovSofteq) [reply redisinsight/api/src/modules/database-recommendation/providers/database-recommendation.provider.ts:166]: Good point

File: redisinsight/api/src/modules/database-recommendation/providers/database-recommendation.provider.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (AmirAllayarovSofteq) [reply redisinsight/api/src/modules/database-recommendation/providers/database-recommendation.provider.ts]: Yep, but others will be in same order...

**Pull request merged**